design/httpproxy-status-conditions.md

```diff
@@ -0,0 +1,305 @@
+
+
+## Goals
+
+- To define a standard Condition schema for HTTPProxy and its sibling CRDs (currently TLSDelegation, but should be applicable to any new CRDs also).
+- To add a place in the CRDs for the Conditions to live.
+
+## Non-goals
+Add support for some method of exposing the inclusion graph in the HTTPProxy’s status. That is a separate issue (#xxx)
+
+## High-level design
+
+We're going to do the following things:
+- add a `conditions` section to HTTPProxy `status`. We're doing this to allow for additional extensibility in the standard way.
+- Add a `Valid` condition as the standard HTTPProxy condition. We've chosen a `Valid` condition here rather than a more traditional `Ready` because `Ready` implies that we're sure that the Envoy config has been accepted - which we don't have the mechanisms to ensure yet. We may add a `Ready` condition at a later date.
+- The `Valid` condition's `Reason` field and the `validity` field will have the same value, as will the `Valid` condition's message field and the `description` field. That is, we will keep the functionality of the `validity` and `description` fields, but also duplicate it in the `Valid` Condition.
+- Add `warnings` and `errors` as special sets of sub-conditions under the `Valid` condition. We need a way to provide additional detail to `Valid: false` HTTPProxies, and we also need a way to surface warnings. Tim Hockin suggested sub-conditions as a possible way of handling this, and it seems like this may be a good fit for our particular use case.
+- The TLSDelegation `status` will also have a `conditions` section, with a `Valid` condition as above.
+
+The `conditions` stanza also allows for external controllers to add information to the HTTPProxy (in particular, this allows `external-dns` to possibly add a `DNSProvisioned` condition or similar in the future).
+
+Yes, the overloading of the term `conditions` between `status` and `spec` is unfortunate, but in order to be able to use standard tooling, we are stuck with the YAML key name for `status`.
+
+## Detailed design
+
+We will create the following generic type as a stand-in until the `metav1.Condition` type is available upstream.
+
+```go
+type Condition struct {
+	// Type of condition in CamelCase or in foo.example.com/CamelCase.
+	// Many .condition.type values are consistent across resources like Available, but because arbitrary conditions can be
+	// useful (see .node.status.conditions), the ability to deconflict is important.
+	// +required
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
```

Comment:
Recommend that (here and below), we use a string enum for the Type and Reason fields. This is transparent on the wire, but gives is a structure to hang documentation from.

Reply:
It does mean that other people can't add additional Types into the `DetailedCondition` list though, which I would argue is desirable. We can document the Contour-supported values here, but making it an enum breaks extensibility - that's the reason for Condtions in the first place.

Reply:
IIRC the argument for other controllers manipulating conditions is that they can (relatively blindly) add their own conditions. If you need to change a condition that another controller creates, then you need to understand its semantics.

Reply:
Yeah, but if the `type` field is an enum, don't they then need to be able to add something to the Contour repo to be able to add their own type value?

Reply:
If you use a Go string type, there's no wire enforcement, so it would be fine AFAIK. It would only be a constraint if you added a kubebuilder validation comment.